staticd/static_nb_config.c

```diff
@@ -0,0 +1,1150 @@
+				   const struct lyd_node *rn_dnode,
+				   struct stable_info *info)
+{
+	struct route_node *rn;
+	struct static_path *pi;
+	const struct lyd_node *pi_dnode;
+	struct static_nexthop *nh;
+	int ret;
+
+	nh = nb_running_unset_entry(args->dnode);
+	pi_dnode = yang_dnode_get_parent(args->dnode, "path-list");
+	pi = nb_running_get_entry(pi_dnode, NULL, true);
+	rn = nb_running_get_entry(rn_dnode, NULL, true);
+
+	ret = static_delete_nexthop(rn, pi, info->safi, info->svrf, nh);
+	if (!ret) {
+		char buf[SRCDEST2STR_BUFFER];
+
+		flog_warn(EC_LIB_NB_CB_CONFIG_APPLY,
+			  "%s : nh [%d:%s:%s:%s] nexthop destroy failed",
+			  srcdest_rnode2str(rn, buf, sizeof(buf)),
+			  yang_dnode_get_enum(args->dnode, "./nh-type"),
+			  yang_dnode_get_string(args->dnode, "./interface"),
+			  yang_dnode_get_string(args->dnode, "./gateway"),
+			  yang_dnode_get_string(args->dnode, "./vrf"));
+		return NB_ERR;
+	}
+
+	return NB_OK;
+}
+
+static int nexthop_mpls_label_stack_entry_create(struct nb_cb_create_args *args)
+{
+	struct static_nexthop *nh;
+	uint8_t id;
+
+	switch (args->event) {
+	case NB_EV_VALIDATE:
+	case NB_EV_PREPARE:
+	case NB_EV_ABORT:
+		break;
+	case NB_EV_APPLY:
+		nh = nb_running_get_entry(args->dnode, NULL, true);
+		id = yang_get_list_pos(args->dnode->parent) - 1;
+		nh->snh_label.label[id] = 0;
+		nh->snh_label.num_labels++;
+		break;
+	}
+
+	return NB_OK;
+}
+
+static int
+nexthop_mpls_label_stack_entry_destroy(struct nb_cb_destroy_args *args)
+{
+	struct static_nexthop *nh;
+	uint8_t id;
+
+	switch (args->event) {
+	case NB_EV_VALIDATE:
+	case NB_EV_PREPARE:
+	case NB_EV_ABORT:
+		break;
+	case NB_EV_APPLY:
+		nh = nb_running_get_entry(args->dnode, NULL, true);
+		id = yang_get_list_pos(args->dnode->parent) - 1;
+		nh->snh_label.label[id] = 0;
+		nh->snh_label.num_labels--;
+		break;
+	}
+
+	return NB_OK;
+}
+
+static int static_nexthop_mpls_label_modify(struct nb_cb_modify_args *args)
+{
+	struct static_nexthop *nh;
+	uint8_t id;
+
+	nh = nb_running_get_entry(args->dnode, NULL, true);
+	/* Mapping to array = list-index -1 */
+	id = yang_get_list_pos(args->dnode->parent) - 1;
+	nh->snh_label.label[id] = yang_dnode_get_uint32(args->dnode, NULL);
+
+	return NB_OK;
+}
+
+static int static_nexthop_mpls_label_destroy(struct nb_cb_destroy_args *args)
+{
+	struct static_nexthop *nh;
+	uint8_t id;
+
+	nh = nb_running_get_entry(args->dnode, NULL, true);
+	/* Mapping to array = list-index -1 */
+	id = yang_get_list_pos(args->dnode->parent);
+	nh->snh_label.label[id] = 0;
+
+	return NB_OK;
+}
+
+static int static_nexthop_onlink_modify(struct nb_cb_modify_args *args)
+{
+	struct static_nexthop *nh;
+
+	nh = nb_running_get_entry(args->dnode, NULL, true);
+	nh->onlink = yang_dnode_get_bool(args->dnode, NULL);
+
+	return NB_OK;
+}
+
+static int static_nexthop_onlink_destroy(struct nb_cb_destroy_args *args)
+{
+	struct static_nexthop *nh;
+
+	nh = nb_running_get_entry(args->dnode, NULL, true);
+	nh->onlink = false;
```

Comment:
Indeed.

Reply:
done, removed the destroy callback.

Reply:
@vishaldhingra You're right. We could change the order of the checks in the NB to avoid the need for a `destroy()` callback in such cases. But I don't know, in some exceptional cases it might be necessary to do something when a leaf with a default value ceases to exist (when it's "when" check changes from true to false). So for the moment let's keep things this way.